### PR TITLE
Eeitt 866 - Repeating sections, new approach

### PR DESCRIPTION
**Pull request by tenochdlc**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala [hunk 1]
```diff
@@ -53,6 +54,24 @@ object FormTemplateValidator {
...
+      (section.repeatsMax, section.repeatsMin) match {
+        case (Some(repMax), Some(repMin)) if !repMax.equals(repMin) =>
+          Invalid(s"The repeatsMax and repeatsMin fields must be the same in a repeating section: repeatsMax=[$repMax], repeatsMin=[$repMin]")
+        case (Some(_), None) | (None, Some(_)) =>
+          val repMax = section.repeatsMax.getOrElse("")
+          val repMin = section.repeatsMin.getOrElse("")
+          Invalid(s"Both repeatsMax and repeatsMin fields must be provided in a repeating section: repeatsMax=[$repMax], repeatsMin=[$repMin]")
+        case _ => Valid
+      }
+    }.partition(_.isInstanceOf[Invalid]) match {
```
Comment (pawelpanasewicz) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala]: This partitioning could be nicer done using 'Monoid.combineAll()'. Have a look:

```
import cats.data._
import cats.implicits._
import cats.kernel.Monoid


val vs: Seq[Validated[String, Int]] = List(1.valid, "err".invalid)

val v: Validated[String, Int] = Monoid.combineAll(vs)
\\ v: cats.data.Validated[String,Int] = Invalid(err)

```

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala [hunk 2]
```diff
@@ -53,6 +54,24 @@ object FormTemplateValidator {
     }
   }
 
+  def validateRepeatingSectionFields(sectionList: List[Section]): ValidationResult = {
+    sectionList.map { section =>
+      (section.repeatsMax, section.repeatsMin) match {
+        case (Some(repMax), Some(repMin)) if !repMax.equals(repMin) =>
```
Comment (pawelpanasewicz) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala]: This condition caught my eye. Why the only valid combination is repMax the same as repMin ?

File: app/uk/gov/hmrc/gform/formtemplate/RepeatingComponentService.scala [hunk 3]
```diff
@@ -81,23 +81,42 @@ object RepeatingComponentService {
...
+      shortName = buildText(section.shortName, index, data),
       fields = section.fields.map(copyField)
     )
   }
 
-  private def buildText(template: Option[String], index: Int, fieldToTrack: VariableInContext,
-    data: Map[String, String]): Option[String] = {
+  private def buildText(template: Option[String], index: Int, data: Map[String, String]): Option[String] = {
+
+    def evaluateTextExpression(str: String) = {
```
Comment (pawelpanasewicz) [inline app/uk/gov/hmrc/gform/formtemplate/RepeatingComponentService.scala]: This code looks very magic for me. Is there a way to make it more understandable?

File: app/uk/gov/hmrc/gform/formtemplate/RepeatingComponentService.scala [hunk 4]
```diff
@@ -81,23 +81,42 @@ object RepeatingComponentService {
...
+      val field = str.replaceFirst("""\$\{""", "").replaceFirst("""\}""", "")
+      if (field.startsWith("n_")) {
+        if (index == 1) {
+          val fieldName = field.replaceFirst("n_", "")
+          data.getOrElse(fieldName, "")
```
Comment (pawelpanasewicz) [inline app/uk/gov/hmrc/gform/formtemplate/RepeatingComponentService.scala]: `data.getOrElse(fieldName, "")` - this worries me. We result in default value when there is a missing field. In my opinion this should be exceptional situation when field can't be resolved and reported as an Exception. Let's discuss it. @lutnos

File: app/uk/gov/hmrc/gform/formtemplate/RepeatingComponentService.scala [hunk 5]
```diff
@@ -81,23 +81,42 @@ object RepeatingComponentService {
...
+        }
+      } else {
+        data.getOrElse(field, "")
+      }
+    }
 
-    val fieldName = if (index == 1) fieldToTrack.field else s"${index - 1}_${fieldToTrack.field}"
-    val textToInsert = data.getOrElse(fieldName, "")
+    def getEvaluatedText(str: String) = {
```
Comment (pawelpanasewicz) [inline app/uk/gov/hmrc/gform/formtemplate/RepeatingComponentService.scala]: How about moving this part and above to separate object (let's say `ExpressionEvaluator`)? How do we do expression evaluation in other places? Maybe what I'm saying here doesn't make sense.

File: test/uk/gov/hmrc/gform/services/RepeatingComponentServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+import uk.gov.hmrc.gform.formtemplate.RepeatingComponentService
+import uk.gov.hmrc.gform.sharedmodel.UserId
+import uk.gov.hmrc.gform.sharedmodel.form._
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+
+import scala.collection.immutable.List
+
+class RepeatingComponentServiceSpec extends Spec {
+
+  val basicFormTemplate = FormTemplate(
```
Comment (pawelpanasewicz) [inline test/uk/gov/hmrc/gform/services/RepeatingComponentServiceSpec.scala]: As previously maybe this is a good place to benefit from `ExampleData`?

File: app/uk/gov/hmrc/gform/core/parsers/ValueParser.scala [hunk 7]
```diff
@@ -30,8 +30,10 @@ object ValueParser {
 
   def validate(expression: String): Opt[ValueExpr] = validateWithParser(expression, exprDeterminer)
 
+  /*
```
Comment (tenochdlc) [inline app/uk/gov/hmrc/gform/core/parsers/ValueParser.scala]: Deleted

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala (hunk 2 above)
Comment (tenochdlc) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala]: Ask Nathan  :-)

**Pull request merged**